README.md
# turkish-llm
playground to create turkish llms

# Leaderboard

| Models     | max-len  | ntok      | nchar      | nvoc   | sum_nll       | sum_bits     | nll/tok | nll/char | token_ppl | char_ppl | bits/tok | bits/char |
|------------|----------|-----------|------------|--------|---------------|--------------|---------|----------|-----------|----------|----------|-----------|
| llama2-7b  |    4096  | 6,672,520 | 13,488,318 | 32,000 |  14,809,664.0 | 21,365,828.0 | 2.2195  | 1.097962 | 9.2027    | 2.9980   | 3.2021   | 1.5840    |
| mistral-7b |    4096  | 6,679,977 | 13,488,318 | 32,000 |  11,988,665.0 | 17,295,988.0 | 1.7653  | 0.888818 | 5.8438    | 2.4322   | 2.5469   | 1.2822    |
| vicuna-7b  |    4096  | 6,672,520 | 13,488,318 | 32,000 |  14,571,667.0 | 21,022,472.0 | 2.1838  | 1.080317 | 8.8802    | 2.9456   | 3.1506   | 1.5585    |
| llama2-7b  |    2048  | 6,672,520 | 13,488,318 | 32,000 |  13,262,011.0 | 19,133,038.0 | 1.9875  | 0.983222 | 7.2976    | 2.6730   | 2.8674   | 1.4184    |
| mistral-7b |    2048  | 6,679,977 | 13,488,318 | 32,000 |  12,210,814.0 | 17,616,480.0 | 1.7980  | 0.905288 | 6.0381    | 2.4726   | 2.5941   | 1.3060    |
| vicuna-7b  |    2048  | 6,672,520 | 13,488,318 | 32,000 |  14,809,664.0 | 21,365,828.0 | 2.2195  | 1.097962 | 9.2027    | 2.9980   | 3.2020   | 1.5840    |
| falcon-7b  |    2048  | 6,486,017 | 13,488,318 | 65,024 |  15,143,924.0 | 21,848,064.0 | 2.3348  | 1.122743 | 10.327    | 3.0732   | 3.3684   | 1.6197    |
| phi-1.5    |    2048  | 6,730,586 | 13,488,318 | 51,200 |  26,347,170.0 | 38,010,932.0 | 3.9145  | 1.953332 | 50.126    | 7.0521   | 5.6474   | 2.8180    |
| pythia-6.9b|    2048  | 6,149,190 | 13,488,318 | 50,432 |  12,318,955.0 | 17,772,496.0 | 2.0033  | 0.913305 | 7.4138    | 2.4925   | 2.8902   | 1.3176    |
| opt-6.7b   |    2048  | 6,735,965 | 13,488,318 | 50,272 |  12,028,472.0 | 17,353,416.0 | 1.7857  | 0.891769 | 5.9638    | 2.4394   | 2.5762   | 1.2865    |
| mpt-7b     |    1024  | 6,149,190 | 13,488,318 | 50,432 |  13,496,818.0 | 19,471,792.0 | 2.1948  | 1.000630 | 8.9790    | 2.7199   | 3.1665   | 1.4436    |
| gpt2-xl    |    1024  | 6,730,965 | 13,488,318 | 50,257 |  23,702,484.0 | 34,195,456.0 | 3.5214  | 1.757260 | 33.832    | 5.7965   | 5.0803   | 2.5351    |


| Models     | max-len  | ntok      | nchar      | nvoc   | sum_nll       | sum_bits     | nll/tok | nll/char | token_ppl | char_ppl | bits/tok | bits/char |
|------------|----------|-----------|------------|--------|---------------|--------------|---------|----------|-----------|----------|----------|-----------|
| xglm-7.5b  |    2048  | 3,308,314 | 13,488,318 | 256,008|   8,505,928.0 | 12,271,460.0 | 2.5710  | 0.630614 | 13.079    | 1.8787   | 3.7092   | 0.9097    |
| xglm-4.5b  |    2048  | 3,308,314 | 13,488,318 | 256,008|   9,073,295.0 | 13,089,998.0 | 2.7425  | 0.672677 | 15.526    | 1.9594   | 3.9566   | 0.9704    |
| xglm-2.9b  |    2048  | 3,308,314 | 13,488,318 | 256,008|   8,970,696.0 | 12,941,978.0 | 2.7115  | 0.665071 | 15.052    | 1.9446   | 3.9119   | 0.9594    |
| xglm-1.7b  |    2048  | 3,308,314 | 13,488,318 | 256,008|   9,517,353.0 | 13,730,638.0 | 2.8767  | 0.705599 | 17.757    | 2.0250   | 4.1503   | 1.0179    |
| xglm-564M  |    2048  | 3,308,314 | 13,488,318 | 256,008|  10,704,995.0 | 15,444,043.0 | 3.2357  | 0.793649 | 25.426    | 2.2114   | 4.6682   | 1.1449    |
| mGPT       |    2048  | 4,393,160 | 13,488,318 | 100,000|  21,218,566.0 | 30,611,920.0 | 4.8299  | 1.573106 | 125.19    | 4.8216   | 6.9680   | 2.2695    |
| bloomz-7b1 |    2048  | 5,459,180 | 13,488,318 | 250,880|  20,515,212.0 | 29,597,194.0 | 3.7579  | 1.520961 | 42.859    | 4.5766   | 5.4215   | 2.1942    |
| blz-7b1-mt |    2048  | 5,459,180 | 13,488,318 | 250,880|  20,547,892.0 | 29,644,342.0 | 3.7639  | 1.523384 | 43.116    | 4.5877   | 5.4301   | 2.1977    |

### Calendar
- [x] step1: Select a (i) monolingual and (ii) multilingual model. 
- [x] step2: Create alapaca-tr data.
- [ ] step3: Select a pretraining dataset.
- [ ] step4: lora-tune (ii) on step2.
- [ ] step5: lora-train (i) on step3 and then step2.
- [ ] step6: create rlaif dataset.
- [ ] step7: eval everything on turkish benchmarks.

ChangeLog
2023-11-28
        *128xparquet data has 129,486,207,634 train tokens and 6,720,970 tokens val tokens
        * We can start trainings for each method but first understand how many iterations to do. How we calculate it?

        TODO:
        - Train scratch GPT-125m on 128xparquet with 8xGPU. 
        - LoRA train Mistral with 1x, 5x, 10x parquet data.
        - LoRA train XGLM-7.5b with 1x, 5x, 10x parquet data.


2023-11-11
        * nanoGPT script added and running successfully for 1xGPU. Lets see for multi-gpu settings.
        * Partial data is loadded, need to load all data.
        * transfer the repo to Truba.


2023-10-30
        * 4000 token is fine max-length according to distributions.
        * lets start with full-parameter: 
                1. train Mistral with 0.5Gb 1Gb, 2Gb, 5Gb, 11Gb as CLM.
                2. fine-tune Mistral alpaca-lora for each.
                3. train XGLM with 0.5Gb 1Gb, 2Gb, 5Gb, 11Gb as CLM.
                4. fine-tune XGLM alpaca-lora for each.
        * Do lora training: 
                1. train Mistral with 0.5Gb 1Gb, 2Gb, 5Gb, 11Gb as CLM.
                2. fine-tune Mistral alpaca-lora for each.
                3. train XGLM with 0.5Gb 1Gb, 2Gb, 5Gb, 11Gb as CLM.
                4. fine-tune XGLM alpaca-lora for each.

2023-10-25
        * Alpaca-tr is finalized with gpt-3.5-instruct and costs ~33$ for 51,658 data.
        * There are noisy examples and generally biased to national question, towards to Turkey(?).
        * Need to find a high quality pre-training split; Culturax is one opportunity.
        * Ask oflazer to private script is another choice.
        * But how many of them we should use it for pre-training?
        * Are we going to do curriculm learning: (i) base-model => pretraining => sft => rlaif, (ii) multilingual-model => sft => rlaif, (iii) multilingual-model => pretraining => sft => rlaif?

2023-10-20
        * Multi-lingual and Mono-lingual experiments finalized; Mistral-7b and XGLM-7.5b is the winners.
        * Time to write the code scripts for alpaca-data generation, get seeds from doga and mete.
        * The cost can be too much, make cost approximation both for (i) gpt-3.5-instruct and (ii) gpt-4.
        * Decide pre-training data for Mono-lingual model; Culturax-175GB?
        * Next week finalize the datasets both for pre-training and instruction-tuning.
        
        TODO:
        - Generate alpaca data from seed-tr.
        - Create pre-training and instruction-tuning splits.

2023-10-19  
        * Opt and mistral gives the best ppl for monolingual settings.
        * Now try xglm-7.5b, mGPT, flan-T5-xxl, bloomz.

2023-10-18

        * We should evaluate the ppl of current LLMs to see their Turkish knowledge.
        * Two options: train multilingual-LLM or teach monolingual-LLM with further CLM.
        * Start getting by bpc for monolingual models on tr-news: llama2-7b, mistral-7b, falcon-7b, vicuna-7b, mpt-7b, phi-1.5, pythia, gpt2, opt.
        * Try different max-windows: 4096, 2048, 1024, 512, 256, 128.
        
        TODO:
        - Try multilingual models: xglm-564M, xglm-1.7B, xglm-2.9B, xglm-4.5B, xglm-7.5B, mGPT, flan-T5, bloomz.
        - Try other turkish-llms all in huggingface.


pretraining/train.py
"""
This training script can be run both on a single gpu in debug mode,
and also in a larger training run with distributed data parallel (ddp).

To run on a single GPU, example:
$ python train.py --batch_size=32 --compile=False

To run with DDP on 4 gpus on 1 node, example:
$ torchrun --standalone --nproc_per_node=4 train.py

To run with DDP on 4 gpus across 2 nodes, example:
- Run on the first (master) node with example IP 123.456.123.456:
$ torchrun --nproc_per_node=8 --nnodes=2 --node_rank=0 --master_addr=123.456.123.456 --master_port=1234 train.py
- Run on the worker node:
$ torchrun --nproc_per_node=8 --nnodes=2 --node_rank=1 --master_addr=123.456.123.456 --master_port=1234 train.py
(If your cluster does not have Infiniband interconnect prepend NCCL_IB_DISABLE=1)
"""

import os
import time
import pickle
from contextlib import nullcontext

import fire
import numpy as np
import torch
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.distributed import init_process_group, destroy_process_group

from model import GPTConfig, GPT
from utils import get_batch, estimate_loss, get_lr


def train(
    # I/O
    out_dir: str = 'out',
    eval_interval: int = 2000,
    log_interval: int = 1,
    eval_iters: int = 200,
    eval_only: bool = False,               # if True, script exits right after the first eval
    always_save_checkpoint: bool = True,   # if True, always save a checkpoint after each eval
    init_from: str = 'scratch',            # 'scratch' or 'resume' or 'gpt2*'
    # wandb logging
    wandb_log: bool = True,               # disabled by default
    wandb_project: str = 'llm-tr',
    wandb_run_name: str = 'gpt2',          # 'run' + str(time.time())
    # data
    dataset: str = 'uonlp/CulturaX',
    gradient_accumulation_steps: int = 40, # 5 * 8, used to simulate larger batch sizes
    batch_size = 12,                       # if gradient_accumulation_steps > 1, this is the micro-batch size
    block_size = 1024,
    # model
    n_layer: int = 12,
    n_head: int = 12,
    n_embd: int = 768,
    dropout: float = 0.0,                  # for pretraining 0 is good, for finetuning try 0.1+
    bias: bool = False,                    # do we use bias inside LayerNorm and Linear layers?
    # adamw optimizer
    learning_rate: float = 6e-4,           # max learning rate
    max_iters: int = 600000,               # total number of training iterations
    weight_decay: float = 1e-1,
    beta1: float = 0.9,
    beta2: float = 0.95,
    grad_clip: float = 1.0,                # clip gradients at this value, or disable if == 0.0
    # learning rate decay settings
    decay_lr: bool = True,                 # whether to decay the learning rate
    warmup_iters: int = 2000,              # how many steps to warm up for
    lr_decay_iters: int = 600000,          # should be ~= max_iters per Chinchilla
    min_lr: float = 6e-5,                  # minimum learning rate, should be ~= learning_rate/10 per Chinchilla
    # DDP settings
    backend: str = 'nccl',                 # 'nccl', 'gloo', etc.
    # system
    device: str = 'cuda',                  # examples: 'cpu', 'cuda', 'cuda:0', 'cuda:1' etc., or try 'mps' on macbooks
    compile: bool = True,                  # use PyTorch 2.0 to compile the model to be faster
):

    # various inits, derived attributes, I/O setup
    config = locals()
    dtype = 'bfloat16' if torch.cuda.is_available() and torch.cuda.is_bf16_supported() else 'float16' # 'float32', 'bfloat16', or 'float16', the latter will auto implement a GradScaler
    ddp = int(os.environ.get('RANK', -1)) != -1 # is this a ddp run?
    config['ddp'] = ddp
    config['dtype'] = dtype
    print(f"config: {config}\n")

    if ddp:
        init_process_group(backend=backend)
        ddp_rank = int(os.environ['RANK'])
        ddp_local_rank = int(os.environ['LOCAL_RANK'])
        ddp_world_size = int(os.environ['WORLD_SIZE'])
        device = f'cuda:{ddp_local_rank}'
        torch.cuda.set_device(device)
        master_process = ddp_rank == 0 # this process will do logging, checkpointing etc.
        seed_offset = ddp_rank # each process gets a different seed
        # world_size number of processes will be training simultaneously, so we can scale
        # down the desired gradient accumulation iterations per process proportionally
        assert gradient_accumulation_steps % ddp_world_size == 0
        gradient_accumulation_steps //= ddp_world_size
    else:
        # if not ddp, we are running on a single gpu, and one process
        master_process = True
        seed_offset = 0
        ddp_world_size = 1
    tokens_per_iter = gradient_accumulation_steps * ddp_world_size * batch_size * block_size
    print(f"tokens per iteration will be: {tokens_per_iter:,}")

    if master_process:
        os.makedirs(out_dir, exist_ok=True)
    torch.manual_seed(1337 + seed_offset)
    torch.backends.cuda.matmul.allow_tf32 = True # allow tf32 on matmul
    torch.backends.cudnn.allow_tf32 = True # allow tf32 on cudnn
    device_type = 'cuda' if 'cuda' in device else 'cpu' # for later use in torch.autocast
    # note: float16 data type will automatically use a GradScaler
    ptdtype = {'float32': torch.float32, 'bfloat16': torch.bfloat16, 'float16': torch.float16}[dtype]
    ctx = nullcontext() if device_type == 'cpu' else torch.amp.autocast(device_type=device_type, dtype=ptdtype)

    # poor man's data loader
    data_dir = os.path.join('data', dataset)
    train_data = np.memmap(os.path.join(data_dir, 'train-parquet-all.bin'), dtype=np.uint16, mode='r')
    val_data = np.memmap(os.path.join(data_dir, 'val-parquet-all.bin'), dtype=np.uint16, mode='r')
    

    # init these up here, can override if init_from='resume' (i.e. from a checkpoint)
    iter_num = 0
    best_val_loss = 1e9

    # attempt to derive vocab_size from the dataset
    meta_path = os.path.join(data_dir, 'meta.pkl')
    meta_vocab_size = None
    if os.path.exists(meta_path):
        with open(meta_path, 'rb') as f:
            meta = pickle.load(f)
        meta_vocab_size = meta['vocab_size']
        print(f"found vocab_size = {meta_vocab_size} (inside {meta_path})")

    # model init
    model_args = dict(n_layer=n_layer, n_head=n_head, n_embd=n_embd, block_size=block_size, bias=bias, vocab_size=None, dropout=dropout) # start with model_args from command line
    if init_from == 'scratch':
        # init a new model from scratch
        print("Initializing a new model from scratch")
        # determine the vocab size we'll use for from-scratch training
        if meta_vocab_size is None:
            print("defaulting to vocab_size of GPT-2 to 50304 (50257 rounded up for efficiency)")
        model_args['vocab_size'] = meta_vocab_size if meta_vocab_size is not None else 50304
        gptconf = GPTConfig(**model_args)
        model = GPT(gptconf)
    elif init_from == 'resume':
        print(f"Resuming training from {out_dir}")
        # resume training from a checkpoint.
        ckpt_path = os.path.join(out_dir, 'ckpt.pt')
        checkpoint = torch.load(ckpt_path, map_location=device)
        checkpoint_model_args = checkpoint['model_args']
        # force these config attributes to be equal otherwise we can't even resume training
        # the rest of the attributes (e.g. dropout) can stay as desired from command line
        for k in ['n_layer', 'n_head', 'n_embd', 'block_size', 'bias', 'vocab_size']:
            model_args[k] = checkpoint_model_args[k]
        # create the model
        gptconf = GPTConfig(**model_args)
        model = GPT(gptconf)
        state_dict = checkpoint['model']
        # fix the keys of the state dictionary :(
        # honestly no idea how checkpoints sometimes get this prefix, have to debug more
        unwanted_prefix = '_orig_mod.'
        for k,v in list(state_dict.items()):
            if k.startswith(unwanted_prefix):
                state_dict[k[len(unwanted_prefix):]] = state_dict.pop(k)
        model.load_state_dict(state_dict)
        iter_num = checkpoint['iter_num']
        best_val_loss = checkpoint['best_val_loss']
    elif init_from.startswith('gpt2'):
        print(f"Initializing from OpenAI GPT-2 weights: {init_from}")
        # initialize from OpenAI GPT-2 weights
        override_args = dict(dropout=dropout)
        model = GPT.from_pretrained(init_from, override_args)
        # read off the created config params, so we can store them into checkpoint correctly
        for k in ['n_layer', 'n_head', 'n_embd', 'block_size', 'bias', 'vocab_size']:
            model_args[k] = getattr(model.config, k)
    # crop down the model block size if desired, using model surgery
    if block_size < model.config.block_size:
        model.crop_block_size(block_size)
        model_args['block_size'] = block_size # so that the checkpoint will have the right value
    model.to(device)

    # initialize a GradScaler. If enabled=False scaler is a no-op
    scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))

    # optimizer
    optimizer = model.configure_optimizers(weight_decay, learning_rate, (beta1, beta2), device_type)
    if init_from == 'resume':
        optimizer.load_state_dict(checkpoint['optimizer'])
    checkpoint = None # free up memory

    # compile the model
    if compile:
        print("compiling the model... (takes a ~minute)")
        unoptimized_model = model
        model = torch.compile(model) # requires PyTorch 2.0

    # wrap model into DDP container
    if ddp:
        model = DDP(model, device_ids=[ddp_local_rank])

    # logging
    if wandb_log and master_process:
        import wandb
        wandb.init(project=wandb_project, name=wandb_run_name, config=config, entity='kuisai', dir="/localscratch/eacikgoz/")

    # training loop
    X, Y = get_batch(
        split='train', 
        train_data=train_data, 
        val_data=val_data, 
        block_size=block_size, 
        batch_size=batch_size, 
        device_type=device_type, 
        device=device,
        ) # fetch the very first batch
    t0 = time.time()
    local_iter_num = 0 # number of iterations in the lifetime of this process
    raw_model = model.module if ddp else model # unwrap DDP container if needed
    running_mfu = -1.0
    while True:

        # determine and set the learning rate for this iteration
        lr = get_lr(
            iter_num=iter_num,
            warmup_iters=warmup_iters, 
            learning_rate=learning_rate, 
            lr_decay_iters=lr_decay_iters, 
            min_lr=min_lr
        ) if decay_lr else learning_rate
        for param_group in optimizer.param_groups:
            param_group['lr'] = lr

        # evaluate the loss on train/val sets and write checkpoints
        if iter_num % eval_interval == 0 and master_process:
            losses = estimate_loss(
                model=model, 
                eval_iters=eval_iters, 
                ctx=ctx, 
                train_data=train_data, 
                val_data=val_data, 
                block_size=block_size, 
                batch_size=batch_size, 
                device_type=device_type, 
                device=device,
            )
            print(f"step {iter_num}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")
            if wandb_log:
                wandb.log({
                    "iter": iter_num,
                    "train/loss": losses['train'],
                    "val/loss": losses['val'],
                    "lr": lr,
                    "mfu": running_mfu*100, # convert to percentage
                })
            if losses['val'] < best_val_loss or always_save_checkpoint:
                best_val_loss = losses['val']
                if iter_num > 0:
                    checkpoint = {
                        'model': raw_model.state_dict(),
                        'optimizer': optimizer.state_dict(),
                        'model_args': model_args,
                        'iter_num': iter_num,
                        'best_val_loss': best_val_loss,
                        'config': config,
                    }
                    print(f"saving checkpoint to {out_dir}")
                    torch.save(checkpoint, os.path.join(out_dir, 'ckpt.pt'))
        if iter_num == 0 and eval_only:
            break

        # forward backward update, with optional gradient accumulation to simulate larger batch size
        # and using the GradScaler if data type is float16
        for micro_step in range(gradient_accumulation_steps):
            if ddp:
                # in DDP training we only need to sync gradients at the last micro step.
                # the official way to do this is with model.no_sync() context manager, but
                # I really dislike that this bloats the code and forces us to repeat code
                # looking at the source of that context manager, it just toggles this variable
                model.require_backward_grad_sync = (micro_step == gradient_accumulation_steps - 1)
            with ctx:
                logits, loss = model(X, Y)
                loss = loss / gradient_accumulation_steps # scale the loss to account for gradient accumulation
            # immediately async prefetch next batch while model is doing the forward pass on the GPU
            X, Y = get_batch(
                split='train', 
                train_data=train_data, 
                val_data=val_data, 
                block_size=block_size, 
                batch_size=batch_size, 
                device_type=device_type, 
                device=device,
            )
            # backward pass, with gradient scaling if training in fp16
            scaler.scale(loss).backward()
        # clip the gradient
        if grad_clip != 0.0:
            scaler.unscale_(optimizer)
            torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip)
        # step the optimizer and scaler if training in fp16
        scaler.step(optimizer)
        scaler.update()
        # flush the gradients as soon as we can, no need for this memory anymore
        optimizer.zero_grad(set_to_none=True)

        # timing and logging
        t1 = time.time()
        dt = t1 - t0
        t0 = t1
        if iter_num % log_interval == 0 and master_process:
            # get loss as float. note: this is a CPU-GPU sync point
            # scale up to undo the division above, approximating the true total loss (exact would have been a sum)
            lossf = loss.item() * gradient_accumulation_steps
            if local_iter_num >= 5: # let the training loop settle a bit
                mfu = raw_model.estimate_mfu(batch_size * gradient_accumulation_steps, dt)
                running_mfu = mfu if running_mfu == -1.0 else 0.9*running_mfu + 0.1*mfu
            print(f"iter {iter_num}: loss {lossf:.4f}, time {dt*1000:.2f}ms, mfu {running_mfu*100:.2f}%")
        iter_num += 1
        local_iter_num += 1

        # termination conditions
        if iter_num > max_iters:
            break

    if ddp:
        destroy_process_group()


if __name__ == "__main__":
    fire.Fire(train)